internal/controller/nodestate_cleaner_controller.go

```diff
@@ -0,0 +1,133 @@
+)
+
+const (
+	// PodNameAnnotation is the annotation key used to store the pod name in FRRNodeState.
+	PodNameAnnotation = "frrk8s.metallb.io/pod-name"
+)
+
+// NodeStateCleaner reconciles Pod objects to clean up FRRNodeState resources.
+type NodeStateCleaner struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Logger log.Logger
+}
+
+// +kubebuilder:rbac:groups=frrk8s.metallb.io,resources=frrnodestates,verbs=get;list;watch;delete
+
+func (r *NodeStateCleaner) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	level.Info(r.Logger).Log("controller", "NodeStateCleaner", "start reconcile", req.String())
+	defer level.Info(r.Logger).Log("controller", "NodeStateCleaner", "end reconcile", req.String())
+
+	frrNodeStates := &frrk8sv1beta1.FRRNodeStateList{}
+	if err := r.Client.List(ctx, frrNodeStates); err != nil {
+		level.Error(r.Logger).Log("controller", "NodeStateCleaner", "failed to list FRRNodeStates", err)
+		return ctrl.Result{}, err
+	}
+
+	var errors []error
+	for _, nodeState := range frrNodeStates.Items {
+		podName, exists := nodeState.Annotations[PodNameAnnotation]
+		if !exists {
+			level.Info(r.Logger).Log("controller", "NodeStateCleaner", "FRRNodeState has no pod name annotation", "name", nodeState.Name)
+			continue
+		}
+		if podName == "" {
+			level.Info(r.Logger).Log("controller", "NodeStateCleaner", "FRRNodeState has empty pod name annotation", "name", nodeState.Name)
+			continue
+		}
+
+		pod := &corev1.Pod{}
+		err := r.Client.Get(ctx, types.NamespacedName{Name: podName, Namespace: req.Namespace}, pod)
+		if k8serrors.IsNotFound(err) {
+			level.Info(r.Logger).Log("controller", "NodeStateCleaner", "deleting FRRNodeState", "name", nodeState.Name, "pod", podName)
+			if err := r.Client.Delete(ctx, &nodeState); err != nil {
+				level.Error(r.Logger).Log("controller", "NodeStateCleaner", "failed to delete FRRNodeState", "name", nodeState.Name, "error", err)
+				errors = append(errors, fmt.Errorf("failed to delete FRRNodeState %s: %w", nodeState.Name, err))
+			}
+			continue
+		}
+		if err != nil {
+			level.Error(r.Logger).Log("controller", "NodeStateCleaner", "failed to get pod", "pod", podName, "error", err)
+			errors = append(errors, fmt.Errorf("failed to get pod %s: %w", podName, err))
+		}
+	}
+
+	if len(errors) > 0 {
+		level.Error(r.Logger).Log("controller", "NodeStateCleaner", "reconcile finished with errors", "error_count", len(errors))
+		return ctrl.Result{}, fmt.Errorf("reconcile finished with %d errors", len(errors))
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *NodeStateCleaner) SetupWithManager(mgr ctrl.Manager) error {
+	startupChan := make(chan event.GenericEvent, 1)
+	// Trigger initial reconciliation, as pod might have
+	// been deleted while we are down.
+	go func() {
+		startupChan <- NewStartupEvent()
+	}()
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Pod{}).
```

Comment:
it _might_ be a stretch, but we could also watch create/deletes of nodestates (that way we also keep the relation between pod<->state e.g if a user creates a state for no reason, and we can drop the startupChan which is a bit ugly)